### DISCUSSION TITLE
Mill config file concept (RFC)

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#1226] [Ideas]
Some settings are relevant for mill but can not be controlled via the `build.sc`. Thus these should be placed into a config file.

I think we should have only one config file which holds all various settings, that are needed to control mills behavior. In contrast to having multiple files which would clutter the directories. As such a single config file need to serve multiple entry points (e.g. the mill executable (JVM), the various start scripts (shell script), metals, IDEs) we should choose a very simple file format, which does not require any sophisticated parsing. I think some property file style without any logic should serve that purpose very well. We should also migrate and deprecate other supported config files like `.mill-version` after that. 

These config files should be project specific as well as user specific, so we need to look for it in the project directory (highest priority) as well as in the user config directory (fallback).

The following settings are suspected to be configurable:

* JVM settings
** JVM client settings (memory, stack, ...) for mill client
** JVM server settings (memory, stack, ...) for mill server
* mill version (replaces `.mill-version`)
* mill default settings (job count, colors (#471), verbose, no-server)

We need a good name for it, my candidates are:
* `.millrc` - this is very Unix/Linux style, I'm used to it but don't know if some users expect some specific format
* `.mill.conf` - this is probably the clearest choice
* `.mill.properties` - this implies JVM properties files in ISO 8859-1 format, where I would prefer UTF-8 content

Example:

```conf
# comments start with a # and will be ignored
mill.version=0.8.0
mill.jobs=0
mill.server.jvm.opts=-Xms2g -Xmx2g -Xss4m -XX:ReservedCodeCacheSize=512m -XX:MaxMetaspaceSize=512m
mill.client.jvm.opts= 
```

What do you think?

**Comment by lihaoyi** [comment]
This seems like a good idea to me

**Comment by lefou** [comment]
We might also consider the following addition settings:
* proxy settings
* Ivy/Maven/Coursier repository settings
* Environment variables

**Comment by joan38** [comment]
+1 for `.mill.conf`

**Comment by Ichoran** [comment]
Why can't these things be controlled via build.sc?  If you can't do it with normal syntax, add a special header.

```
/*!! mill.conf {
  "mill.version" = "0.8.0",
 "mill.server.jvm.opts" = [
    "-Xms2g", "-Xmx2g", "-Xss4m",
    "-XX:ReservedCodeCacheSize=512m",
    "-XX:MaxMetaspaceSize=512m"
  ],
  "mill.client.jvm.opts" = []
} !!*/
```

(Or use the variant you showed above instead of JSON.)

**Comment by lefou** [comment]
It could, but the effort to parse it somehow from the `*.sc` file will be far from trivial, especially for a small shell script, which needs to decide which version of mill to download. Also you have to ensure no editor ever messes up with that comment. We should keep things simple, if possible.

**Comment by Ichoran** [comment]
That isn't hard if you make the start and stop lines really easy to match.  And whatever parses the build.sc can throw an error if they're malformed (if it will even run at all).

See several solutions at https://stackoverflow.com/questions/4857424/extract-lines-between-2-tokens-in-a-text-file-using-bash

**Comment by joan38** [comment]
I found out the we currently have `.mill-jvm-opts` and want to document that here so that we eventually merge that in the new config system.

**Comment by nrktkt** [comment]
Seems like a good idea. 

One detail that occurs to me is that it might be awkward to have mill server options configurable at the project level. We could a) have those settings only available at the user level or b) restart the server every time the project changes (this does happen already for mill version changes I think).

**Comment by lefou** [comment]
Experience has shown, that you always need some kind of project-specific environment settings for your tool chain. Each project has different needs, e.g. memory requirements. The JVM defaults highly depend on the user environment like OS, count of CPU cores, RAM, and so on. Project-specific version-controlled settings ensure a reproducible setup for team members, CI systems and your future self. The mill server is the component which holds the various tools in caches, so it's the mill server runtime settings which determine the available memory for your compiler.

**Comment by nafg** [comment]
There could potentially be 3 categories (maybe not all populated): user-level-only settings, project-level-only settings, and settings that can be configured at both levels, project-level taking precedence.

Some examples from this thread that would make the most sense to be user-level: proxy settings, colors

Jobs is actually interesting, because you might want to have one setting locally for a project and a different one in CI. For example CI often has far fewer vCPUs and less RAM so it's more efficient and safer to have fewer threads. (More efficient because less context switching and safer because the fewer things you do at a time the less memory you need at a time. For example running scalajs on a few modules at a time has caused me OOMs on CI with too much currency, at least with SBT.) But I guess that's what the command line flag is for. :wink:

**Comment by lefou** [comment]
Additional to-be configurable properties:
* max count of parallel mill server instances (mill-worker) per project (currently hardcoded)
* location where mill should store/search for server instance state (mill-worker), default to `out/mill-worker-*`, but in cases where `out`is on a mounted / remote / in-memory drive, in may be desirable to override this (currently hardcoded)

**Reply by lefou** [reply]
This whole config file idea requires us the refactor settings handling in Mill. Currently, most settings only have a configuration point in the CLI parser config. After we implemented this idea, we may have various levels of config files. This gives us the chance to also support environment variables for some settings. Also, the CLI parser needs to be aware of variable defaults or of the fact whether an options was set or not. 

We may end up with the following configuration priority list:

* (system config - e.g. `/etc/mill.conf`)
* user config - `$XDG_CONFIG_DIR/mill/mill.conf` or `$HOME/.mill.conf`
* project config `./.mill.conf`
* Environment variables
* Mill CLI options

**Comment by nafg** [comment]
That sounds good, and standard (e.g. git has the same priority list)

On Fri Apr 1, 2022, 11:09 AM GMT, Tobias Roeser ***@***.***>
wrote:

This whole config file idea requires us the refactor settings handling in
Mill. Currently, most settings only have a configuration point in the CLI
parser config. After we implemented this idea, we may have various levels
of config files. This gives us the change to also support environment
variables for some settings. Also, the CLI parser needs to be aware of
variable defaults or of the fact whether an options was set or not.

We may end up with the following configuration priority list:

   - (system config - e.g. /etc/mill.conf)
   - user config - $XDG_CONFIG_DIR/mill/mill.conf or $HOME/.mill.conf
   - project config ./.mill.conf
   - Environment variables
   - Mill CLI options

—
Reply to this email directly, view it on GitHub
<https://github.com/com-lihaoyi/mill/discussions/1226#discussioncomment-2485088>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AAAYAUGV7XIWPZP45MX5WPTVC3KOFANCNFSM5MLHFVNA>
.
You are receiving this because you commented.Message ID:
***@***.***>

**Comment by DamianReeves** [comment]
> we should choose a very simple file format, which does not require any sophisticated parsing

If there was some sort of native launcher for mill (let's call it `milln`) would we be amicable to a config format which did not have to meet the requirement that it is parseable by a shell script?

I am playing around with scala native and seeing if I can make a serviceable `millw` equivalent that is written in Scala and is meant only to exist for the Scala NAtive platform.

**Reply by lefou** [reply]
IMHO, as long as the structure of our configuration can be kept simple, we should also keep the file format simple. If you have a good case for some more complex structure, please present it here, so we can reevaluate.

**Reply by DamianReeves** [reply]
Well I was thinking we could use `metaconfig` like [scalafmt](https://github.com/scalameta/scalafmt/blob/master/build.sbt#L136-L141) and [scalafix](https://github.com/scalacenter/scalafix/blob/main/build.sbt#L73-L74) do. This would allow us to use HOCON configs.

**Reply by lefou** [reply]
I think the config example from the OP can be parsed with typesafe config library.

**Comment by lefou** [comment]
> Why can't these things be controlled via build.sc? If you can't do it with normal syntax, add a special header.
> 
> ```
> /*!! mill.conf {
>   "mill.version" = "0.8.0",
>  "mill.server.jvm.opts" = [
>     "-Xms2g", "-Xmx2g", "-Xss4m",
>     "-XX:ReservedCodeCacheSize=512m",
>     "-XX:MaxMetaspaceSize=512m"
>   ],
>   "mill.client.jvm.opts" = []
> } !!*/
> ```
> 
> (Or use the variant you showed above instead of JSON.)

While thinking about a solution to 

* https://github.com/com-lihaoyi/mill/issues/1675

I came across the using directive, which is used by Scala-cli and available as a stand-alone Java library (https://github.com/VirtusLab/using_directives), and which could be a lightweight replacement for Ammonite magic imports (`$`-imports). If we decide to use it and proof of concepts work, we could as well think of supporting some of these settings directly via using directives.

```scala
//> using dep "some.thirdparty::dependency::0.1.0"
//> using file "ci/some-script.sc"
//> using mill.version 0.10.5
```

**Comment by jackcviers** [comment]
These are all already possible via the default `scalaConfFile` pulled in by `Resolve` in coursier.

By default,  it uses the [scala-cli config file](https://scala-cli.virtuslab.org/docs/commands/config/).

That [config file](https://scala-cli.virtuslab.org/docs/reference/commands#config) supports many of the options suggested in this thread:

- proxy configuration 
- credentials configuration 
- publishing and signing keys configuration 
- github token configuration 

It also has key value parsers and secrets value parsers built-in, and is already included in mill's dependencies. 

`Resolve` takes options for multiple config and property files already, and also handles ones passed via jvm property or environment variable. 

`Resolve#finalRepositories` uses the config files already,  but it's implementation is broken (it tosses out the configuration file repository info by flattening incorrectly, applies mirror configuration but not credentials). It can, of course, be fixed upstream or we could use our own PlatformResolve implementation to override and extend its current behavior for the ideas in this thread. 

To maintain backward compatibility, I suggest adding an `ExternalConfig` trait users can mix into their builds, which overrides the `CoursierModule#repositoriesTask` to use the default config correctly while an upstream fix can be negotiated, and to dogfood the changes presented in this thread. 

I've already been working on using this config functionality that is already built-in for managing our mill builds at work. I'd love to contribute this.

**Reply by lefou** [reply]
Thanks for this input.

> `Resolve#finalRepositories` uses the config files already, but it's implementation is broken (it tosses out the configuration file repository info by flattening incorrectly, applies mirror configuration but not credentials). It can, of course, be fixed upstream or we could use our own PlatformResolve implementation to override and extend its current behavior for the ideas in this thread.

Do you have any pointers? An open issue, maybe?

**Reply by jackcviers** [reply]
I'm in the middle of making a PR to coursier. I've already tested the change to ensure it is loading the config properly and it is working:

```diff
modified   modules/coursier/shared/src/main/scala/coursier/Resolve.scala
@@ -61,13 +61,7 @@ import scala.language.higherKinds
   def finalRepositories: F[Seq[Repository]] = {
     val repositories0 =
       if (preferConfFileDefaultRepositories) {
-        val defaultFromConfOpt = confFiles
-          .iterator
-          .flatMap(Resolve.confFileRepositories(_).iterator)
-          .take(1)
-          .toList
-          .headOption
-        defaultFromConfOpt.getOrElse(repositories)
+        confFiles.flatMap(Resolve.confFileRepositories(_).getOrElse(repositories)).toList
       }
       else
         repositories
```

I still need to apply the authentication to the imported repositories. This is binary-backwards-compatible, and source-compatible so should result in a patch version in coursier, and then we can simply upgrade mill's dependencies.

For the additional config options we are talking about here, we can use the same `CoursierPaths` `Resolve` uses to load the default configuration, and/or provide additional config paths in our own `MillPaths` modeled after `CoursierPaths` (it uses the proper overloading of envVar > local > correct user home preferences location for the configs) to load individual mill configs we want to set things through, then we can implement the key parsers, probably using the same parsers the cli uses, and then we can actually probably just glue it in to where we set the values now.

This would be binary-backwards-compatible, as we aren't changing any existing interfaces, just adding new ones and using them in the appropriate places. That means a minor version bump for mill, which means old mill extensions will work just fine with the changes.

When I'm done getting repos to load appropriately, I'll check that the proxy settings in coursier propagate to mill appropriately, and if not, figure out why and fix them in a similar manner to the config above, then I'll start a draft PR for this RFC using the config, resolve, and coursier paths + possible mill paths to get things rolling here.

Alternatively, while I'm handling coursier repo config, someone else could handle the draft pr for using coursier paths to get this rolling, and I could come jump over and help after all that is done. I also want to port sbt-github-actions to mill for my new company, which would probably take about a week or so.

Work can always intercede. I'll let you know here when the pr is up over on the coursier repo. I'll also need to make an issue repository so that I can report the issue and attach it to the pr. I'm trying to finish all this up today if possible.

**Reply by jackcviers** [reply]
An update -- looks like Mill uses the default `FilesCache` from coursier, which should pick up the credentials from the default config file. What is missing in `CoursierModule` is setting up the proxy for coursier, which is done with `Resolve#setupProxy`. I'll make a follow up PR to ensure this is done in mill.

**Reply by jackcviers** [reply]
An update;

So the default config _does_ work in coursier! Became clear once I wrote the tests.

It also gets pulled in via mill `repositoryTask`. What doesn't work is that while coursier attaches the credentials when downloading the assets, apparently mill requires them on the repositories themselves (and in fact turns off the credentials coursier uses in its `FileCache`.

So now I believe the thing to do is to attach the credentials to the resolved repositories in `CoursierModule#repositoryTask` like this:

```scala
Ok. so credentials and default repositories actually do work from the coursier cli! So my little change wasn't needed. However, `mill` resolves things slightly differently than coursier does.

coursier doesn't require me to define the authentication on the repositories it parses when it resolves dependencies from an authenticated repository. However, `mill` requires the authentication on the repositories themselves. I'm not sure why.

However, overriding the repositoryTask in a build to look like this:

```scala
import scala.concurrent.ExecutionContext.Implicits.global
val resolvedCredentials = coursier.cache.CacheDefaults.credentials.flatMap{_.get()}
val repos = Await.result(
    Resolve().finalRepositories.map{_.map{
      case x:IvyRepository =>
        x.withAuthentication(
          resolvedCredentials.find{ c =>
            c.matches(x.pattern.string, c.usernameOpt.getOrElse(""))
          }.map(_.authentication))
      case x:MavenRepository =>
        x.withAuthentication(
          resolvedCredentials.find{ c =>
            c.matches(x.root, c.usernameOpt.getOrElse(""))
          }.map(_.authentication))
      }}.future(),
    Duration.Inf
)
repos
```

This would then use the coursier credentials properties file/environment variable/the default scala config file for any credential definition, use the coursier repositories environment variable/the default scala config file to define the values, and to add proxy we can change the above to:

```scala
val resolve = Resolve.proxySetup()
...
```

So I'll code that up and attach this discussion to the PR.

**Reply by jackcviers** [reply]
PR here: https://github.com/com-lihaoyi/mill/pull/2954

**Comment by lefou** [comment]
The following PR adds support for reading the preferred Mill version from a using directive in the `build.mill` file.

* https://github.com/com-lihaoyi/mill/pull/4624
